src/lightning/data/utilities/broadcast.py

```diff
@@ -0,0 +1,187 @@
+        message = f"HTTPClient: Method: {func.__name__.upper()}, Path: {args[0]}\n"
+        message += f"      Base URL: {self.base_url}\n"
+        params = kwargs.get("query_params", {})
+        if params:
+            message += f"      Params: {params}\n"
+        resp: requests.Response = func(self, *args, **kwargs)
+        message += f"      Response: {resp.status_code} {resp.reason}"
+        self.log_function(message)
+        return resp
+
+    return wrapped
+
+
+def _response(r: Any, *args: Any, **kwargs: Any) -> Any:
+    return r.raise_for_status()
+
+
+class HTTPClient:
+    """A wrapper class around the requests library which handles chores like logging, retries, and timeouts
+    automatically."""
+
+    def __init__(
+        self,
+        base_url: str,
+        auth_token: Optional[str] = None,
+        log_callback: Optional[Callable] = None,
+        use_retry: bool = True,
+    ) -> None:
+        self.base_url = base_url
+        retry_strategy = Retry(
+            # wait time between retries increases exponentially according to: backoff_factor * (2 ** (retry - 1))
+            # but the the maximum wait time is 120 secs. By setting a large value (2880), we'll make sure clients
+            # are going to be alive for a very long time (~ 4 days) but retries every 120 seconds
+            total=_CONNECTION_RETRY_TOTAL,
+            backoff_factor=_CONNECTION_RETRY_BACKOFF_FACTOR,
+            status_forcelist=[
+                408,  # Request Timeout
+                429,  # Too Many Requests
+                500,  # Internal Server Error
+                502,  # Bad Gateway
+                503,  # Service Unavailable
+                504,  # Gateway Timeout
+            ],
+        )
+        adapter = CustomRetryAdapter(max_retries=retry_strategy, timeout=_DEFAULT_REQUEST_TIMEOUT)
+        self.session = requests.Session()
+
+        self.session.hooks = {"response": _response}
+
+        if use_retry:
+            self.session.mount("http://", adapter)
+            self.session.mount("https://", adapter)
+
+        if auth_token:
+            self.session.headers.update({"Authorization": f"Bearer {auth_token}"})
+
+        self.log_function = log_callback or self.log_function  # type: ignore
+
+    @_http_method_logger_wrapper
+    def get(self, path: str) -> Any:
+        url = urljoin(self.base_url, path)
+        return self.session.get(url)
+
+    @_http_method_logger_wrapper
+    def post(
+        self, path: str, *, query_params: Optional[Dict] = None, data: Optional[bytes] = None, json: Any = None
+    ) -> Any:
+        url = urljoin(self.base_url, path)
+        return self.session.post(url, data=data, params=query_params, json=json)
+
+    @_http_method_logger_wrapper
+    def delete(self, path: str) -> Any:
+        url = urljoin(self.base_url, path)
+        return self.session.delete(url)
+
+    def log_function(self, message: str, *args: Any, **kwargs: Any) -> None:
```

Comment:
What is the example for it though? Are we going to use it ourselves?

Reply:
This isn't needed. I removed it.